fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/services/ReplacePropertiesServiceImpl.java

```diff
@@ -56,35 +58,63 @@ public void perform(final Transaction tx,
-                .triples(fromModel(inputModel.createResource(fedoraId.getFullId()).asNode(), inputModel))
-                .build();
+            final var rdfStream = fromModel(inputModel.createResource(fedoraId.getFullId()).asNode(), inputModel);
+            final var serverManagedMode = fedoraPropsConfig.getServerManagedPropsMode();
+
+            // create 2 updates -- one for the properties coming in and one for and server managed properties
+            final ResourceOperation primaryOp;
+            final Optional<ResourceOperation> secondaryOp;
+            if (fedoraId.isDescription()) {
+                primaryOp = factory.updateBuilder(tx, fedoraId, serverManagedMode)
+                                   .userPrincipal(userPrincipal)
+                                   .triples(rdfStream)
+                                   .build();
+
+                // we need to use the description id until we write the headers in order to resolve properties
+                secondaryOp = Optional.of(factory.updateManagedHeadersBuilder(tx, fedoraId, serverManagedMode)
+                                                 .relaxedProperties(inputModel)
+                                                 .userPrincipal(userPrincipal)
+                                                 .build());
+            } else {
+                primaryOp = factory.updateBuilder(tx, fedoraId, serverManagedMode)
+                                   .relaxedProperties(inputModel)
+                                   .userPrincipal(userPrincipal)
+                                   .triples(rdfStream)
+                                   .build();
+                secondaryOp = Optional.empty();
+            }
 
             lockArchivalGroupResource(tx, pSession, fedoraId);
             tx.lockResource(fedoraId);
             if (RdfLexicon.FEDORA_NON_RDF_SOURCE_DESCRIPTION_URI.equals(interactionModel)) {
                 tx.lockResource(fedoraId.asBaseId());
             }
 
-            pSession.persist(updateOp);
+            pSession.persist(primaryOp);
             updateReferences(tx, fedoraId, userPrincipal, inputModel);
             membershipService.resourceModified(tx, fedoraId);
             searchIndex.addUpdateIndex(tx, pSession.getHeaders(fedoraId, null));
-            recordEvent(tx, fedoraId, updateOp);
+            recordEvent(tx, fedoraId, primaryOp);
+            secondaryOp.ifPresent(operation -> updateBinaryHeaders(tx, pSession, operation));
         } catch (final PersistentStorageException ex) {
             throw new RepositoryRuntimeException(String.format("failed to replace resource %s",
                     fedoraId), ex);
         }
     }
+
+    private void updateBinaryHeaders(final Transaction tx,
+                                     final PersistentStorageSession pSession,
+                                     final ResourceOperation operation) {
+        tx.lockResource(operation.getResourceId());
```

Comment:
I don't think you need this lock. I believe the resource is already locked on line 98